include/lin/references.hpp

```diff
@@ -1,11 +1,450 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+
 /** @file lin/references.hpp
  *  @author Kyle Krol
- *  Collection of header files composing the references module. */
+ */
+
+/** @defgroup REFERENCES References
+ *
+ *  @brief Allows interpretting portions of larger tensor as other tensor types.
+ *
+ *  This can be extremely helpful if, for example, you'd like to normalize all
+ *  the columns of a matrix without copying elements in and out of the matrix
+ *  itself. We'll use this an example demonstrating how to use the refernces
+ *  module.
+ *
+ *  First, be sure to include the appropriate headers (that being the
+ *  `%lin/core.hpp` and `%lin/references.hpp` headers). From there, we can
+ *  implement the following function which will normalize the columns in a three
+ *  by three matrix:
+ *
+ *  ~~~{.cpp}
+ *  #include <lin/core.hpp>
+ *  #include <lin/references.hpp>
+ *
+ *  template <typename T>
+ *  void normalize(lin::Matrix<T, 3, 3> &M) {
+ *    for (lin::size_t j = 0; j < M.cols(); j++) {
+ *      auto m = lin::ref<lin::Vector<T, 3>>(M, 0, j);
+ *      m = m / lin::norm(m);
+ *    }
+ *  }
+ *  ~~~
+ *
+ *  where line of interest here is clearly the call to
+ *  `%lin::ref<%lin::Vector<T, 3>>` which essentially asks for a reference to be
+ *  created that acts like a `%lin::Vector<T, 3>` and whose first element
+ *  matches up with `M(0, j)`.
+ *
+ *  There are a couple things to note here:
+ *
+ *   - The reference itself is invalid when the backing object (in this case
+ *     `M`) goes out of scope.
+ *   - This particular reference serves as a read and write interface. However,
+ *     read only references can also be generated in the same fashion is the
+ *     underlying type is `const` or read only.
+ *
+ *  Lastly, there are also convenience functions `lin::ref_col` and
+ *  `lin::ref_row` to more easily reference the entire column or row of a tensor
+ *  directly. In fact, if you don't mind using the `lin::internal` namespace a
+ *  little, a simple function normalizing the columns of any matrix can be
+ *  implemented as shown below:
+ *
+ *  ~~~{.cpp}
+ *  #include <lin/core.hpp>
+ *  #include <lin/references.hpp>
+ *
+ *  template <class D>
+ *  void normalize(lin::internal::Mapping<D> &M) {
+ *    for (lin::size_t j = 0; j < M.cols(); j++) {
+ *      auto m = lin::ref_col(M, j);
+ *      m = m / lin::norm(m);
+ *    }
+ *  }
+ *  ~~~
```

Comment:
how would you do this without ref_col or ref_row, and what's the backing of a matrix?

Is it row vectors or col vectors or something else?

Reply:
The "backing matrix/tensor" refers to the `lin` type a reference uses to produce and assign elements values. If I write something like this:

    lin::Matrix3x3d A;
    auto a = lin::ref<lin::Vector3d>(A, 0, 0);

the "backing matrix" to reference `a` is `A` - i.e. whenever you ask for an elements of `a` like `a(0)` it will access `A` (in this case) `A(0, 0)` to produce it.

Reply:
Writing that function without `ref_col` would basically require you to implement the `ref_col` function within the function itself. You'd need to pull the traits from the templated type `D` like so:

    template <class D, typename = std::enable_if_t<lin::internal::disjunction<
        lin::internal::is_matrix<D>, lin::internal::is_col_vector<D>>::value>>
    void normalize(lin::internal::Mapping<D> &M) {
      typedef typename D::Traits::elem_t elem_t;
      constexpr lin::size_t rows = D::Traits::rows;
      constexpr lin::size_t max_rows = D::Traits::max_rows;

      for (lin::size_t j = 0; j < M.cols(); j++) {
        auto m = lin::ref<lin::Vector<elem_t, rows, max_rows>>(M, 0, j);
        m = m / lin::norm(m);
      }
    }